### PR TITLE
mlx - implement segment_sum and segment_max

### PR DESCRIPTION
**Pull request by lkarthee**

### CODE REVIEW

File: keras/src/backend/mlx/math.py [hunk 1]
```diff
@@ -6,11 +6,37 @@
 
 
 def segment_sum(data, segment_ids, num_segments=None, sorted=False):
-    raise NotImplementedError("segment_sum is not implemented for mlx")
+    data = convert_to_tensor(data)
```
Comment (lkarthee) [inline keras/src/backend/mlx/math.py:9]: mlx does not have unique and boolean indices  https://github.com/ml-explore/mlx/issues/246  and https://github.com/ml-explore/mlx/issues/377 . So the implementation is buggy with varying values of num_segments and negative segment_ids.

Comment (lkarthee) [reply keras/src/backend/mlx/math.py:9]: One way to solve this is using numpy.unique on segment_ids to determine num_segments.

**Comment by Faisal-Alsrheed** [comment]
Hi @lkarthee,

I have noticed the same issue: there are missing and needed functions in MLX. 

What is the best strategy here?  Should we use NumPy temporarily and add a TODO comment?

I have decided to start with code files and functions. 
https://github.com/keras-team/keras/pull/19619

I suggest creating a roadmap where we start with fundamental functions and then move up.

best

**Comment by lkarthee** [comment]
@Faisal-Alsrheed My thoughts regarding missing functions which can't be added due to design decisions/limitations of mlx: 
- we have to check if there is a workaround. 
- if there is no workaround, may be we have to fallback to using numpy or jax  adding a TODO?

Regarding fixing core first, I have been doing that in my PRs. I tried to implement Pooling and CNN related funcs, realised many tests are failing and started fixing test cases. 

@fchollet any thoughts on what to use as fallback.

File: keras/src/backend/mlx/math.py (hunk 1 above)
Comment (fchollet) [reply keras/src/backend/mlx/math.py:9]: @angeloskath could you please advise here?

Comment (gbaned) [reply keras/src/backend/mlx/math.py:9]: Hi @angeloskath Any update on above [comment](https://github.com/keras-team/keras/pull/19652#discussion_r1588632247) from @fchollet ? Please. Thank you!

**Comment by gbaned** [comment]
Hi @fchollet Any update on this PR? Please. Thank you!

**Comment by fchollet** [comment]
> @fchollet any thoughts on what to use as fallback - numpy or jax.

Sorry, just checking this now. I would say that JAX is preferable as fallback since numpy would not be sufficiently performant.

**Comment by gbaned** [comment]
Hi @lkarthee  Any update on this PR? Please. Thank you!

**Pull request closed without merging**